base/compiler/ssair/passes.jl

```diff
@@ -1255,3 +1255,185 @@ function cfg_simplify!(ir::IRCode)
     compact.active_result_bb = length(bb_starts)
     return finish(compact)
 end
+
+# function is_known_fcall(stmt::Expr, @nospecialize(func))
+#     isexpr(stmt, :foreigncall) || return false
+#     s = stmt.args[1]
+#     isa(s, QuoteNode) && (s = s.value)
+#     return s === func
+# end
+
+function is_known_fcall(stmt::Expr, funcs::Vector{Symbol})
+    isexpr(stmt, :foreigncall) || return false
+    s = stmt.args[1]
+    isa(s, QuoteNode) && (s = s.value)
+    # return any(e -> s === e, funcs)
+    return true in map(e -> s === e, funcs)
+end
+
+function is_allocation(stmt::Expr)
+    isexpr(stmt, :foreigncall) || return false
+    s = stmt.args[1]
+    isa(s, QuoteNode) && (s = s.value)
+    return (s === :jl_alloc_array_1d
+         || s === :jl_alloc_array_2d
+         || s ===  :jl_alloc_array_3d
+         || s === :jl_new_array)
+end
+
+function memory_opt!(ir::IRCode)
+    compact = IncrementalCompact(ir, false)
+    uses = IdDict{Int, Vector{Int}}()
+    relevant = IdSet{Int}() # allocations
+    revisit = Int[] # potential targets for a mutating_arrayfreeze drop-in
+    maybecopies = Int[] # calls to maybecopy
+
+    function mark_escape(@nospecialize val)
+        isa(val, SSAValue) || return
+        #println(val.id, " escaped.")
+        val.id in relevant && pop!(relevant, val.id)
+    end
+
+    function mark_use(val, idx)
+        isa(val, SSAValue) || return
+        id = val.id
+        id in relevant || return
+        (haskey(uses, id)) || (uses[id] = Int[])
+        push!(uses[id], idx)
+    end
+
+    for ((_, idx), stmt) in compact
```

Comment:
I'm actually not sure if this linear scan is enough to compute escape information correctly.
For example `mark_use` assumes allocation is already found before the use, but I found it doesn't necessarily hold when there is a loop and `Argument` etc.
And this is why EscapeAnalysis.jl iterates over IR until all the states converge.

Reply:
Agreed - this optimization will be much more correct once we can use EscapeAnalyis.jl escape info